### PR TITLE
[CPP Graph]Enable FFN fusion

### PR DESCRIPTION
**Pull request by intellinjun**
## Type of Change

Add FFN fusion for gptneox-20b，Dolly-v2-3B, Mpt-7B, Starcoder-3B, falcon-7B, Bloom

### CODE REVIEW

File: intel_extension_for_transformers/backends/neural_engine/graph/models/gptneox/gptneox.cpp [hunk 1]
```diff
@@ -34,12 +34,16 @@
 #include "models/model_utils/model_config.h"
 #include "models/model_utils/model_utils.h"
 #include "models/model_utils/util.h"
-
+#define FFN_FUSION 1
```
Comment (airMeng) [inline intel_extension_for_transformers/backends/neural_engine/graph/models/gptneox/gptneox.cpp:37]: format

**Comment by intellinjun** [comment]
https://inteltf-jenk.sh.intel.com/view/nlp-toolkit-validation/job/ITREX-cpp-graph-extension/52/

**Comment by airMeng** [comment]
please update with latest API https://github.com/intel/intel-extension-for-transformers/blob/793629f11a775a05b5c22ee53224020516926eac/intel_extension_for_transformers/llm/runtime/graph/models/gptj/gptj.cpp#L204

**Review by a32543254 — APPROVED** [review]
LGTM

**Comment by intellinjun** [comment]
https://inteltf-jenk.sh.intel.com/view/nlp-toolkit-validation/job/ITREX-cpp-graph-extension/71/

File: intel_extension_for_transformers/llm/runtime/graph/models/gptneox/gptneox.cpp [hunk 2]
```diff
@@ -31,14 +31,23 @@
...
+#include "models/model_utils/model_config.h"
 #include "models/model_utils/model_utils.h"
 #include "models/model_utils/util.h"
 
 // feed-forward network
-struct ne_tensor* gpt_neox_ff(const model_layer& layer, ne_context* ctx0, ne_tensor* inp) {
+struct ne_tensor* gpt_neox_ff(const model_layer& layer, const int batch_size, const int N, ne_context* ctx0, ne_tensor* inp) {
   struct ne_tensor* cur = ne_norm(ctx0, inp);
 
   cur = ne_add(ctx0, ne_mul(ctx0, ne_repeat(ctx0, layer.norm[2], cur), cur), ne_repeat(ctx0, layer.norm[3], cur));
```
Comment (luoyu-intel) [inline intel_extension_for_transformers/llm/runtime/graph/models/gptneox/gptneox.cpp:44]: @yuchengliu1 has implemented non-repeat add, please update with him.

**Review by luoyu-intel — CHANGES_REQUESTED** [review]
remove repeat op.

**Review by kevinintel — APPROVED** [review]
LGTM

File: intel_extension_for_transformers/llm/runtime/graph/models/gptneox/gptneox.cpp (hunk 2 above)
Comment (luoyu-intel) [reply intel_extension_for_transformers/llm/runtime/graph/models/gptneox/gptneox.cpp:44]: move to the next pr.

**Review by luoyu-intel — APPROVED** [review]

**Comment by airMeng** [comment]
> https://inteltf-jenk.sh.intel.com/view/nlp-toolkit-validation/job/ITREX-cpp-graph-extension/71/

failed?

**Comment by intellinjun** [comment]
> > https://inteltf-jenk.sh.intel.com/view/nlp-toolkit-validation/job/ITREX-cpp-graph-extension/71/
> 
> failed?

failed Since yesterday a pr in zhentao modified the command line of inference, causing inference to fail, I retested it, still in queue, I tested it locally and there is no problem

File: intel_extension_for_transformers/llm/runtime/graph/models/bloom/bloom.cpp [hunk 3]
```diff
@@ -218,17 +223,23 @@ static bool bloom_model_eval_internal(model_context& lctx, const model_token* to
...
-                model.layers[il].ffn[2],
-                cur);
-        cur = ne_add(ctx0, ne_repeat(ctx0, model.layers[il].ffn[3], cur), cur);
-
-        cur = ne_gelu(ctx0, cur);
-
-        cur = ne_mul_mat(ctx0,
-                model.layers[il].ffn[4],
-                cur);
+        if (jblas_fusion_FFN_Add_GeLu_f32f32_support(model.layers[il].ffn[0]->data, model.layers[il].ffn[2]->data,
```
Comment (Zhenzhong1) [inline intel_extension_for_transformers/llm/runtime/graph/models/bloom/bloom.cpp:226]: Please check input parameters of  jblas_fusion_FFN_Add_GeLu_f32f32_support for Bloom and GPTNeox. Their FFN[0] has different meanings.

**Pull request merged**